Stripe/Payments/STPPaymentManager.m

```diff
@@ -0,0 +1,413 @@
+
+@property (nonatomic, readonly, strong) id<STPAuthenticationContext> authenticationContext;
+@property (nonatomic, readonly, strong) STPAPIClient *apiClient;
+@property (nonatomic, readonly, strong) STPThreeDSCustomizationSettings *threeDSCustomizationSettings;
+@property (nonatomic, readonly, copy) STPPaymentManagerActionCompletionBlock completion;
+
+@property (nonatomic, nullable) STPPaymentIntent *paymentIntent;
+
+@end
+
+@implementation STPPaymentManagerActionParams
+{
+    dispatch_once_t _threeDS2ServiceOnceToken;
+}
+
+@synthesize threeDS2Service = _threeDS2Service;
+
+- (instancetype)initWithAPIClient:(STPAPIClient *)apiClient
+            authenticationContext:(id<STPAuthenticationContext>)authenticationContext
+     threeDSCustomizationSettings:(STPThreeDSCustomizationSettings *)threeDSCustomizationSettings
+                       completion:(STPPaymentManagerActionCompletionBlock)completion {
+    self = [super init];
+    if (self) {
+        _apiClient = apiClient;
+        _authenticationContext = authenticationContext;
+        _threeDSCustomizationSettings = threeDSCustomizationSettings;
+        _completion = [completion copy];
+    }
+
+    return self;
+}
+
+- (nullable STDSThreeDS2Service *)threeDS2Service {
+    dispatch_once(&_threeDS2ServiceOnceToken, ^{
+        self->_threeDS2Service = [[STDSThreeDS2Service alloc] init];
+        @try {
+            STDSConfigParameters *configParams = [[STDSConfigParameters alloc] initWithStandardParameters];
+            [configParams addParameterNamed:@"kInternalStripeTestingConfigParam" withValue:@"Y"];
+            [self->_threeDS2Service initializeWithConfig:configParams
+                                                  locale:[NSLocale autoupdatingCurrentLocale]
+                                              uiSettings:self->_threeDSCustomizationSettings.uiCustomization];
+        } @catch (NSException *e) {
+            self->_threeDS2Service = nil;
+        }
+
+    });
+
+    return _threeDS2Service;
+}
+
+
+@end
+
+@interface STPPaymentManager () <SFSafariViewControllerDelegate, STPURLCallbackListener, STDSChallengeStatusReceiver>
+{
+     STPPaymentManagerActionParams *_currentAction;
+}
+
+@end
+
+@implementation STPPaymentManager
+
++ (instancetype)sharedManager {
+    static STPPaymentManager *sharedManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedManager = [[STPPaymentManager alloc] init];
+        sharedManager->_apiClient = [STPAPIClient sharedClient];
+    });
+
+    return sharedManager;
+}
+
+- (void)confirmPayment:(STPPaymentIntentParams *)paymentParams
+withAuthenticationContext:(id<STPAuthenticationContext>)authenticationContext
+            completion:(STPPaymentManagerActionCompletionBlock)completion {
+    if (_currentAction != nil) {
+        completion(STPPaymentManagerActionStatusFailed, nil, [NSError errorWithDomain:STPPaymentManagerErrorDomain
```

Comment:
I'll add an assertion, but it will be compiled out for release (so will really only assert for us, which is fine). I don't think I want to throw an exception in released code